File: src/test/java/eu/luminis/sample/OrderAggregateTest.java
package eu.luminis.sample;

import eu.luminis.sample.commands.ChangeShippingAddressCommand;
import eu.luminis.sample.commands.CreateOrderCommand;
import eu.luminis.sample.commands.ShipOrderCommand;
import eu.luminis.sample.events.OrderCreatedEvent;
import eu.luminis.sample.events.OrderShippedEvent;
import eu.luminis.sample.exceptions.AlreadyShippedException;
import eu.luminis.sample.values.Address;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class OrderAggregateTest {
    private final UUID millenniumFalconLegoSet = UUID.randomUUID();
    private final Address campusAddress = new Address("One Infinite Loop", "CA 95014", "Cupertino", "United States");
    private final Address operaHouseAddress = new Address("2 Macquarie Street", "NSW 2000", "Sydney", "Australia");

    @Test
    public void shouldCreateOrder() {
        var event = OrderAggregate.handle(
                new CreateOrderCommand(List.of(millenniumFalconLegoSet), campusAddress)
        );

        assertEquals(campusAddress, event.shippingAddress());
    }

    @Test
    public void shouldFailWhenMissingCreatedEvent() {
        assertThrows(IllegalStateException.class, () ->
                new OrderAggregate(List.of())
        );
    }

    @Test
    public void shouldAllowChangingShippingAddressWhenNotShipped() {
        var order = new OrderAggregate(List.of(
                new OrderCreatedEvent(Instant.now(), List.of(millenniumFalconLegoSet), campusAddress)
        ));

        var event = order.handle(
                new ChangeShippingAddressCommand(operaHouseAddress)
        );

        assertEquals(operaHouseAddress, event.shippingAddress());
    }

    @Test
    public void shouldNotAllowShippingOrderWhenAlreadyShipped() {
        var order = new OrderAggregate(List.of(
                new OrderCreatedEvent(Instant.now(), List.of(millenniumFalconLegoSet), campusAddress),
                new OrderShippedEvent(Instant.now(), "UPS")
        ));

        assertThrows(AlreadyShippedException.class, () ->
                order.handle(new ShipOrderCommand("UPS"))
        );
    }

    @Test
    public void shouldNotAllowChangingShippingOrderAfterBeingShipped() {
        var order = new OrderAggregate(List.of(
                new OrderCreatedEvent(Instant.now(), List.of(millenniumFalconLegoSet), campusAddress),
                new OrderShippedEvent(Instant.now(), "UPS")
        ));

        assertThrows(AlreadyShippedException.class, () ->
                order.handle(new ChangeShippingAddressCommand(operaHouseAddress))
        );
    }
}

File: DomainModel/OrderAggregateTest.cs
using OrderAggregate.Command;
using OrderAggregate.Events;
using OrderAggregate.Exceptions;
using OrderAggregate.Values;

namespace OrderAggregate;

[TestClass]
public class OrderAggregateTest
{
    private readonly Guid _millenniumFalconLegoSet = new Guid();
    private readonly Address _campusAddress = new Address("One Infinite Loop", "CA 95014", "Cupertino", "United States");
    private readonly Address _operaHouseAddress = new Address("2 Macquarie Street", "NSW 2000", "Sydney", "Australia");

    [TestMethod]
    public void ShouldCreateOrder()
    {
        var orderCreatedEvent = OrderAggregate.Handle(
            new CreateOrderCommand([_millenniumFalconLegoSet], _campusAddress)
        );

        Assert.AreEqual(_campusAddress, orderCreatedEvent.ShippingAddress);
    }

    [TestMethod]
    public void ShouldFailWhenMissingCreatedEvent()
    {
        Assert.ThrowsException<InvalidOperationException>(() => new OrderAggregate([]));
    }

    [TestMethod]
    public void ShouldAllowChangingShippingAddressWhenNotShipped()
    {
        var order = new OrderAggregate([new OrderCreatedEvent(DateTime.Now, [_millenniumFalconLegoSet], _campusAddress)]);

        var changeShippingEvent = order.Handle(new ChangeShippingAddressCommand(_operaHouseAddress));

        Assert.AreEqual(_operaHouseAddress, changeShippingEvent.ShippingAddress);
    }

    [TestMethod]
    public void ShouldNotAllowShippingOrderWhenAlreadyShipped()
    {
        var order = new OrderAggregate([new OrderCreatedEvent(DateTime.Now, [_millenniumFalconLegoSet], _campusAddress),
        new OrderShippedEvent(DateTime.Now, "UPS")]);

        Assert.ThrowsException<AlreadyShippedException>(() => order.Handle(new ShipOrderCommand("UPS")));
    }

    [TestMethod]
    public void ShouldNotAllowChangingShippingOrderAfterBeingShipped()
    {
        var order = new OrderAggregate([
                new OrderCreatedEvent(DateTime.Now, [_millenniumFalconLegoSet], _campusAddress),
                new OrderShippedEvent(DateTime.Now, "UPS")
        ]);

        Assert.ThrowsException<AlreadyShippedException>(() => order.Handle(new ChangeShippingAddressCommand(_operaHouseAddress)));
    }
}